### PR TITLE
automated XEN guest function test

### PR DESCRIPTION
**Pull request by Walterkd**
Because in the past we do not have the automation testing for XEN PV, so I implement the automation XEN function test to openQA.

- Verification run: http://10.67.132.101/tests/134

### CODE REVIEW

**Comment by Walterkd** [comment]
@wjn740 @asmorodskyi Please help me review it. Thank you!

File: tests/cpu_bugs/xen_pv.pm [hunk 1]
```diff
@@ -0,0 +1,418 @@
...
+#
+# Copyright © 2019 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: CPU BUGS on Linux kernel check
+# Maintainer: James Wang <jnwang@suse.com>
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: Add an author line, written down your name and e-mail.

File: tests/cpu_bugs/xen_pv.pm [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+#
+# This testsuite for the mitigations control on XEN PV guest.
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: Think about HVM stuff. Might be we could integrate HVM control flow here.

File: tests/cpu_bugs/xen_pv.pm [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+#
+# From xen's oneline doc:
+# http://xenbits.xen.org/docs/4.12-testing/misc/xen-command-line.html#xpti-x86
+# http://xenbits.xen.org/docs/4.12-testing/misc/xen-command-line.html#spec-ctrl-x86
+# http://xenbits.xen.org/docs/4.12-testing/misc/xen-command-line.html#pv-l1tf-x86
+#
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: This part might be could remove as we only test the **GUEST** of Xen.

File: tests/cpu_bugs/xen_pv.pm [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+#                 ||
+#                 ||
+#                 ||
+#                 ||
+#                 ||
+#  _______________||_________________
+# |Remove the cmdline and grub_config.|
+# |And pick the next para=value.|___________________________________________________________
+# |When pick the last key value in a dic, jump to the next para and repeat the process above.|
+#  -----------------------------------------------------------------------------------------
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: The graph is ok. But it seems like need to reformat.

File: tests/cpu_bugs/xen_pv.pm [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+#
+#
+#
+#
+#
+#
+#
+#
+# Put into xpti, spec-ctrl, into this structure. and The testing code will read all them and
+# one by one execution checking.
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: too many blank lines.

File: tests/cpu_bugs/xen_pv.pm [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+    script_run(
+"md /tmp/upload_mitigations; cp " . $Mitigation::syspath . "* /tmp/upload_mitigations; cp /proc/cmdline /tmp/upload_mitigations; lscpu >/tmp/upload_mitigations/cpuinfo; tar -jcvf /tmp/upload_mitigations.tar.bz2 /tmp/upload_mitigations"
+    );
+    script_run("xl info");
+    upload_logs '/tmp/upload_mitigations.tar.bz2';
+    script_run("rm -rf /tmp/upload_mitigations");
+    script_run("rm -rf /tmp/upload_mitigations.tar.bz2");
+    remove_grub_cmdline_settings('xpti=[a-z,]*');
+    remove_grub_cmdline_settings('spec-ctrl=[a-z,-]*');
+    remove_grub_cmdline_settings('pv-l1tf=[a-z,]*');
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: These operation is happen at host, not guest. so If you need clean up Guests, you have to change to ssh_vm_cmd() like functions.

**Comment by Walterkd** [comment]
@wjn740 Modification is done.

File: tests/cpu_bugs/xen_pv.pm [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+            if ($ret ne 0) {
+                record_info('ERROR', "$parameter test is failed.", result => 'fail');
+            }
+            record_info('INFO', "$parameter test is finished.");
+            ssh_vm_cmd("sed -i -e 's/GRUB_CMDLINE_LINUX=\"$parameter\"/GRUB_CMDLINE_LINUX=\"\"/' /etc/default/grub");
+            config_and_reboot();
+        }
+    }
+
+
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: blank lines.

File: tests/cpu_bugs/xen_pv.pm [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+            }
+        },
+    },
+};
+
+
+
+
+
+
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: blank lines.

File: tests/cpu_bugs/xen_pv.pm [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+    script_run("rm -rf /tmp/upload_mitigations");
+    script_run("rm -rf /tmp/upload_mitigations.tar.bz2");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/pti=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/l1tf=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/mds=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/tsx_async_abort=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/spectre_v2=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("sed -i '/GRUB_CMDLINE_LINUX=/s/spectre_v2_user=[a-z,]*/\\ /' /etc/default/grub");
+    ssh_vm_cmd("grub2-mkconfig -o /boot/grub2/grub.cfg");
+    grub_mkconfig;
```
Comment (wjn740) [inline tests/cpu_bugs/xen_pv.pm]: This line seems unused.

File: tests/cpu_bugs/xen_pv.pm [hunk 10]
```diff
@@ -0,0 +1,394 @@
+# SUSE's openQA tests
+#
+# Copyright © 2019 SUSE LLC
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: you should change template ;) it's 2020 already :)

File: tests/cpu_bugs/xen_pv.pm [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+# without any warranty.
+
+# Summary: CPU BUGS on Linux kernel check
+# Maintainer: James Wang <jnwang@suse.com>
+# Author: Di Kong <di.kong@suse.com>
+#
+# This testsuite for the mitigations control on XEN PV guest.
+# And we will integrate HVM control flow here soon.
+#
+# We testing each of parameter. To check the influence of each of them.
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]:
```suggestion
# We testing each parameter. To check the influence each of them.
```
my English skills actually pretty bad , but I have a gut feel that this sounds better ...

File: tests/cpu_bugs/xen_pv.pm [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+                    assert_script_run("xl dmesg | grep -A 10 \"Speculative\"");
+                    print "This unexpection is empty string, skip";
+                }
+
+            }
+        }
+    }
+}
+
+sub do_test {
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: not sure why we need this function , it is only thing what you call inside `run` and you doing it once . So why not just move all this code into `run` and have one function less ? Also not sure why you passing `mitigation_list` as input param, variable is visible in scope of `do_test` so just refer directly to it

File: tests/cpu_bugs/xen_pv.pm [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+          . " --memory=8192"
+          . " --vcpu=8"
+          . " --vnc"
+          . " --events on_reboot=destroy"
+          . " --serial pty", timeout => 3600);
+    script_run("virsh domxml-to-native xen-xl --domain ${vm_domain_name} > ${vm_domain_name}.cfg");
+    script_run("xl create \"${vm_domain_name}.cfg\"");
+    script_run('cd mitigation-testsuite');
+    $vm_ip_addr = script_output(
+        "export password=nots3cr3t; ./get-ip-from-alive-guest.expect"
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]:
```suggestion
        "export password=$testapi::password; ./get-ip-from-alive-guest.expect"
```

File: tests/cpu_bugs/xen_pv.pm [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+            }
+            record_info('INFO', "$parameter test is finished.");
+            ssh_vm_cmd("sed -i -e 's/GRUB_CMDLINE_LINUX=\"$parameter\"/GRUB_CMDLINE_LINUX=\"\"/' /etc/default/grub");
+            config_and_reboot();
+        }
+    }
+}
+
+sub config_and_reboot {
+    #my $timeout = @_;
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: keeping commented code is bad practice please delete this line

File: tests/cpu_bugs/xen_pv.pm [hunk 15]
```diff
@@ -0,0 +1,394 @@
...
+    ssh_vm_cmd("grub2-mkconfig -o /boot/grub2/grub.cfg");
+    ssh_vm_cmd("poweroff");
+    script_run('xl list');
+    script_run('xl list');
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: why you call same command twice ?

File: tests/cpu_bugs/xen_pv.pm [hunk 16]
```diff
@@ -0,0 +1,394 @@
...
+};
+
+sub do_check {
+    my $secnario = shift;
+    my $foo      = $secnario->{default};
+    if ($foo->{expected}) {
+        while (my ($cmd, $lines) = each %{$foo->{expected}}) {
+            foreach my $expected_string (@{$lines}) {
+                if ($expected_string ne "") {
+                    #my $ret = script_run("$cmd | grep \"$expected_string\"");
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: please remove commented code

**Comment by Walterkd** [comment]
@asmorodskyi Thank you for your advice. Already modified.

File: tests/cpu_bugs/xen_pv.pm [hunk 17]
```diff
@@ -0,0 +1,391 @@
...
+    my $cmd = shift;
+    my $ret = script_run("sshpass -p ${qa_password} ssh -qy root\@${vm_ip_addr} $cmd");
+    return $ret;
+}
+
+sub run {
+    my $self = @_;
+    select_console 'root-console';
+    die "platform mistake, This system is not running as Dom0." if script_run("test -d /proc/xen");
+    my ($self, $mitigations_list) = @_;
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: delete this line

File: tests/cpu_bugs/xen_pv.pm [hunk 18]
```diff
@@ -0,0 +1,391 @@
...
+                }
+
+            }
+        }
+    }
+    if ($foo->{unexpected}) {
+        while (my ($cmd, $lines) = each %{$foo->{unexpected}}) {
+            foreach my $unexpected_string (@{$lines}) {
+                if ($unexpected_string ne "") {
+                    #my $ret = script_run("$cmd | grep \"$unexpected_string\"");
```
Comment (asmorodskyi) [inline tests/cpu_bugs/xen_pv.pm]: delete this line

**Comment by asmorodskyi** [comment]
looks good to me . 
but can you please squash "something went wrong" commit into another one 
and also I think it would be good idea to re-run verification run on latest version of code

**Comment by asmorodskyi** [comment]
after this I am ok to merge it

**Comment by Walterkd** [comment]
> after this I am ok to merge it

Yes, I have already squash it. And I am re-running verification test. I will update the link after that.
Thanks a lot.

**Comment by wjn740** [comment]
LGTM.

**Comment by wjn740** [comment]
@alice-suse Would you please help review this?

**Pull request merged**